File: models/Gpt-3.5-turbo/py/code/additional_problems/School_Excursion.py

def find_parent(parent, x):
    if parent[x] == x:
        return x
    parent[x] = find_parent(parent, parent[x])
    return parent[x]

def union_groups(parent, group_size, a, b):
    parent_a = find_parent(parent, a)
    parent_b = find_parent(parent, b)

    if parent_a != parent_b:
        if group_size[parent_a] < group_size[parent_b]:
            parent[parent_a] = parent_b
            group_size[parent_b] += group_size[parent_a]
        else:
            parent[parent_b] = parent_a
            group_size[parent_a] += group_size[parent_b]

def main():
    n, m = map(int, input().split())
    parent = list(range(n+1))
    group_size = [1] * (n+1)

    for _ in range(m):
        a, b = map(int, input().split())
        union_groups(parent, group_size, a, b)

    korkeasaari_count = sum(1 for i in range(1, n+1) if find_parent(parent, i) == i)
    bit_string = ''.join('1' if i <= korkeasaari_count else '0' for i in range(1, n+1))

    print(bit_string)

main()

File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/School_Excursion.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<int> parent;
vector<int> groupSize;

int findParent(int node) {
    if (parent[node] == node)
        return node;

    // Path compression
    return parent[node] = findParent(parent[node]);
}

void unionGroups(int a, int b) {
    int parentA = findParent(a);
    int parentB = findParent(b);

    if (parentA == parentB)
        return;

    // Union by size
    if (groupSize[parentA] < groupSize[parentB]) {
        swap(parentA, parentB);
    }

    parent[parentB] = parentA;
    groupSize[parentA] += groupSize[parentB];
}

int main() {
    int n, m;
    cin >> n >> m;

    parent.resize(n + 1);
    groupSize.resize(n + 1, 1);

    for (int i = 1; i <= n; i++) {
        parent[i] = i;
    }

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        unionGroups(a, b);
    }

    vector<int> result(n + 1, 0);

    for (int i = 1; i <= n; i++) {
        int parentI = findParent(i);
        result[parentI] = groupSize[parentI];
    }

    for (int i = 1; i <= n; i++) {
        cout << (result[i] > 0 ? 1 : 0);
    }
    cout << endl;

    return 0;
}
